pkg/dpm/manager.go

```diff
@@ -3,99 +3,208 @@ package dpm
+	// subset of available signals.
+	glog.V(3).Info("Registering for system signal notifications")
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT)
 
-	// The other important channel is filesystem notification channel, responsible for watching device plugin directory.
+	// The other important channel is filesystem notification channel, responsible for watching
+	// device plugin directory.
+	glog.V(3).Info("Registering for notifications of filesystem changes in device plugin directory")
 	fsWatcher, _ := fsnotify.NewWatcher()
+	defer fsWatcher.Close()
 	fsWatcher.Add(pluginapi.DevicePluginPath)
 
-	dpm := &DevicePluginManager{
-		stopCh:    stopCh,
-		signalCh:  sigs,
-		fsWatcher: fsWatcher,
+	// Create list of running plugins and start Discover method of given lister. This method is
+	// responsible of notifying manager about changes in available plugins.
+	var pluginsMap = make(map[string]devicePlugin)
+	glog.V(3).Info("Starting Discovery on new plugins")
+	pluginsCh := make(chan ResourceLastNamesList)
+	defer close(pluginsCh)
+	go dpm.lister.Discover(pluginsCh)
+
+	// Finally start a loop that will handle messages from opened channels.
+	glog.V(3).Info("Handling incoming signals")
+HandleSignals:
+	for {
+		select {
+		case newPluginsList := <-pluginsCh:
+			glog.V(3).Infof("Received new list of plugins: %s", newPluginsList)
+			dpm.handleNewPlugins(pluginsMap, newPluginsList)
+		case event := <-fsWatcher.Events:
+			if event.Name == pluginapi.KubeletSocket {
+				glog.V(3).Infof("Received kubelet socket event: %s", event)
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					dpm.startPluginServers(pluginsMap)
+				}
+				// TODO: Kubelet doesn't really clean-up it's socket, so this is currently
+				// manual-testing thing. Could we solve Kubelet deaths better?
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					dpm.stopPluginServers(pluginsMap)
+				}
+			}
+		case s := <-signalCh:
+			switch s {
+			case syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT:
+				glog.V(3).Infof("Received signal \"%v\", shutting down", s)
+				dpm.shutDownPlugins(pluginsMap)
+				break HandleSignals
+			}
+		}
+	}
+}
+
+func (dpm *Manager) handleNewPlugins(currentPluginsMap map[string]devicePlugin, newPluginsList ResourceLastNamesList) {
+	var wg sync.WaitGroup
+
+	newPluginsSet := make(map[string]bool)
+	for _, newPluginLastName := range newPluginsList {
+		newPluginsSet[newPluginLastName] = true
 	}
 
-	go dpm.handleSignals()
+	// Add new plugins
+	for newPluginLastName, _ := range newPluginsSet {
+		wg.Add(1)
+		go func() {
+			if _, ok := currentPluginsMap[newPluginLastName]; !ok {
+				glog.V(3).Infof("Adding a new plugin \"%s\"", newPluginLastName)
+				plugin := newDevicePlugin(dpm.lister.GetResourceNamespace(), newPluginLastName, dpm.lister.NewPlugin(newPluginLastName))
+				startUpPlugin(newPluginLastName, plugin)
+				currentPluginsMap[newPluginLastName] = plugin
```

Comment:
We need a mutex around this write, otherwise concurrent map writes might occur.

Reply:
Done